pages/docs/manual/latest/async-await.mdx

```diff
@@ -0,0 +1,338 @@
+    }
+  }
+}
+```
+
+Note how we are essentially catching JS errors the same way as described in our [Exception](exception#catch-rescript-exceptions-from-js) section.
+
+You may unify error and value handling in a single switch as well:
+
+```res
+let authenticate = async () => {
+  raise(Js.Exn.raiseRangeError("Authentication failed."))
+}
+
+let checkAuth = async () => {
+  switch await authenticate() {
+  | _ => Js.log("ok")
+  | exception Js.Exn.Error(e) => 
+    switch Js.Exn.message(e) {
+    | Some(msg) => Js.log("JS error thrown: " ++ msg)
+    | None => Js.log("Some other exception has been thrown")
+    }
+  }
+}
+```
+
+**Important:** When using `await` with a `switch`, always make sure to put the actual await call in the `switch` expression, otherwise your `await` error will not be caught.
+
+## Piping `await` calls
+
+You may want to pipe the result of an `await` call right into another function.
+This can be done by wrapping your `await` calls in a new `{}` closure.
+
+<CodeTab labels={["ReScript", "JS Output"]}>
+
+```res example
+@val external fetchUserMail: string => promise<string> = "GlobalAPI.fetchUserMail"
+
+let fetchData = async () => {
+  let mail = {await fetchUserMail("1234")}->Js.String2.toUpperCase
+  Js.log(`All upper-cased mail: ${mail}`)
+}
+```
+
+```js
+async function fetchData(param) {
+  var mail = (await GlobalAPI.fetchUserMail("1234")).toUpperCase();
+  console.log("All upper-cased mail: " + mail + "");
+}
+```
+
+</CodeTab>
+
+Note how the original closure was removed in the final JS output. No extra allocations!
+
+## Pattern matching on `await` calls
+
+`await` calls are just another kind of expression, so you can use `switch` pattern matching for more complex logic.
+
+<CodeTab labels={["ReScript", "JS Output"]}>
+
+```res example
+@val external fetchUserMail: string => promise<string> = "GlobalAPI.fetchUserMail"
+
+let fetchData = async () => {
+  switch (await fetchUserMail("user1"), await fetchUserMail("user2")) {
+  | (user1Mail, user2Mail) => {
+      Js.log("user 1 mail: " ++ user1Mail)
+      Js.log("user 2 mail: " ++ user2Mail)
+    }
+
+  | exception JsError(err) => Js.log2("Some error occurred", err)
+  }
+}
+```
+
+```js
+async function fetchData(param) {
+  var val;
+  var val$1;
+  try {
+    val = await GlobalAPI.fetchUserMail("user1");
+    val$1 = await GlobalAPI.fetchUserMail("user2");
+  }
+  catch (raw_err){
+    var err = Caml_js_exceptions.internalToOCamlException(raw_err);
+    if (err.RE_EXN_ID === "JsError") {
+      console.log("Some error occurred", err._1);
+      return ;
+    }
+    throw err;
+  }
+  console.log("user 1 mail: " + val);
+  console.log("user 2 mail: " + val$1);
+}
+```
+
+</CodeTab>
+
+## `await` multiple promises
+
+We can utilize the `Js.Promise2` module to handle multiple promises. E.g. let's use `Js.Promise2.all` to wait for multiple promises before continuing the program:
+
+```res
+let pauseReturn = async (value, timeout) => {
```

Comment:
This example is weird as `pauseReturn` returns a `promise<promise<'a>>`.

I think it should be

```rescript
let pauseReturn = (value, timeout) => {
  Js.Promise2.make((~resolve, ~reject) => {
    Js.Global.setTimeout(() => {
      resolve(. value)
    }, timeout)->ignore
  })
}

let logMultipleValues = async () => {
  let promise1 = pauseReturn("value1", 2000)
  let promise2 = pauseReturn("value2", 1200)
  let promise3 = pauseReturn("value3", 500)

  let all = await Js.Promise2.all([promise1, promise2, promise3])

  switch all {
  | [v1, v2, v3] => Js.log(`All values: ${v1}, ${v2}, ${v3}`)
  | _ => Js.log("this should never happen")
  }
}
```

Reply:
done.